network/p2p/p2pnode/libp2pNode_test.go

```diff
@@ -384,6 +384,63 @@ func TestCreateStream_SinglePeerDial(t *testing.T) {
 	require.Equal(t, expectedCreateStreamRetries, createStreamRetries.Load(), fmt.Sprintf("expected %d dial peer retries got %d", expectedCreateStreamRetries, createStreamRetries.Load()))
 }
 
+// TestCreateStream_InboundConnResourceLimit ensures that the setting the resource limit config for PeerDefaultLimits.ConnsInbound restricts the number of inbound
+// connections created from a peer to the configured value.
+func TestCreateStream_InboundConnResourceLimit(t *testing.T) {
+	idProvider := mockmodule.NewIdentityProvider(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	signalerCtx := irrecoverable.NewMockSignalerContext(t, ctx)
+
+	sporkID := unittest.IdentifierFixture()
+
+	uniMgrFactory := p2ptest.UnicastManagerFixtureFactory()
+
+	sender, id1 := p2ptest.NodeFixture(
+		t,
+		sporkID,
+		t.Name(),
+		p2ptest.WithConnectionGater(testutils.NewConnectionGater(idProvider, func(pid peer.ID) error {
+			return nil
+		})),
+		p2ptest.WithDefaultResourceManager(),
+		p2ptest.WithUnicastManagerFactoryFunc(uniMgrFactory),
+		p2ptest.WithCreateStreamRetryDelay(10*time.Millisecond))
+
+	receiver, id2 := p2ptest.NodeFixture(
+		t,
+		sporkID,
+		t.Name(),
+		p2ptest.WithDefaultResourceManager(),
+		p2ptest.WithConnectionGater(testutils.NewConnectionGater(idProvider, func(pid peer.ID) error {
+			return nil
+		})),
+		p2ptest.WithCreateStreamRetryDelay(10*time.Millisecond))
+
+	idProvider.On("ByPeerID", sender.Host().ID()).Return(&id1, true).Maybe()
+	idProvider.On("ByPeerID", receiver.Host().ID()).Return(&id2, true).Maybe()
+
+	p2ptest.StartNodes(t, signalerCtx, []p2p.LibP2PNode{sender, receiver}, 100*time.Millisecond)
+	defer p2ptest.StopNodes(t, []p2p.LibP2PNode{sender, receiver}, cancel, 100*time.Millisecond)
+
+	pInfo, err := utils.PeerAddressInfo(id2)
+	require.NoError(t, err)
+	sender.Host().Peerstore().AddAddrs(pInfo.ID, pInfo.Addrs, peerstore.AddressTTL)
+
+	var wg sync.WaitGroup
+	for i := 0; i < 20; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			_, err = sender.CreateStream(ctx, receiver.Host().ID())
+			require.NoError(t, err)
+		}()
+	}
+
+	unittest.RequireReturnsBefore(t, wg.Wait, 2*time.Second, "could not create streams on time")
+	require.Len(t, receiver.Host().Network().ConnsToPeer(sender.Host().ID()), 1)
```

Comment:
1. Please also ensure that the number of created streams between the two nodes is verified, as it should equal 20.
2. If there is no performance regression on the test, please consider a higher number of concurrent streams, e.g., 50. 
3. Please also include a comment here and in the test's godoc documentation, stating that if this test becomes flaky at this point, it indicates a violation of the single inbound connection guarantee. In such cases, the test should not be quarantined but requires immediate resolution.

Reply:
https://github.com/onflow/flow-go/pull/4022/commits/3f271a4e59e20aecd1e40a9ee60f13e2a9d56867

Reply:
Please use `CountStream` [method](https://github.com/onflow/flow-go/blob/f8600aa8ff922e35b0caf15b3b6142a750cac8c2/network/internal/p2putils/utils.go#L81) instead of counting the streams through `streamCreated`. This method counts the streams at the libp2p level which is more reliable than counting through a counter variable.

Reply:
https://github.com/onflow/flow-go/pull/4022/commits/f9f0ee8b9c1239cc214db100d6972dbba79a79fa